### DISCUSSION TITLE
How to trigger CustomWorkflow based on expiry of SLA field present in one of collection in Cosmos db

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#5269] [Q&A]
Consider there is collection in cosmos db named Case, which has SLAExpiry Time, We have to trigger CustomWorkflow if SLAExpiry field is less than or equal to CurrentDateTime in UTC. How to achieve this scenario?

**Comment by glime-ai** [comment]
To achieve the scenario where you want to trigger a custom workflow in Elsa when the `SLAExpiry` field of a document in a Cosmos DB collection named `Case` is less than or equal to the current UTC date and time, you can follow these steps:

1. **Create a Timer Trigger**: Use a timer trigger in your workflow that periodically checks the `Case` collection in Cosmos DB. The frequency of this check will depend on your requirements for how often you want to check for SLA expiry.

2. **Fetch Documents**: Within the workflow, implement an activity that queries the Cosmos DB `Case` collection to fetch documents where the `SLAExpiry` field is less than or equal to the current UTC date and time.

3. **Filter Documents**: After fetching the documents, filter them to find the ones that meet the SLA expiry condition.

4. **Trigger Custom Workflow**: For each document that meets the condition, trigger the custom workflow. You can use the `IWorkflowRuntime` service to start a workflow instance programmatically.

5. **Handle Workflow Logic**: In the custom workflow, implement the logic that should be executed when an SLA expiry is detected.

Here is a pseudo-code example of how you might set up the initial workflow with a timer trigger:

```csharp
builder
    .StartWith<TimerEvent>(setup => setup.CronExpression = "*/5 * * * *") // Runs every 5 minutes
    .Then<FetchDocumentsFromCosmosDb>() // Custom activity to fetch documents from Cosmos DB
    .Then<FilterExpiredSLADocuments>() // Custom activity to filter documents based on SLAExpiry
    .ForEach(doc => doc.FilteredDocuments)
        .Do(branch => branch
            .StartWith<TriggerCustomWorkflow>(setup => setup.WorkflowDefinitionId = "custom-workflow-id")); // Trigger custom workflow for each document
```

Please note that you will need to implement the `FetchDocumentsFromCosmosDb` and `FilterExpiredSLADocuments` activities according to your specific application logic and Cosmos DB SDK usage.

Remember to handle any potential issues such as network failures, document fetch errors, and ensure that the workflow is idempotent to avoid duplicate processing.

If you need further assistance with the implementation details or have specific questions about the custom activities, please let me know.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
